File: backend/app/schemas/ground_truth.py
"""
SatyaSetu Backend — Ground Truth Schemas (Phase 3)
Pydantic schemas for canonical tenders, requirements, bidders, evidence, and compliance benchmarks.
"""

from typing import Optional, Any
from pydantic import BaseModel, Field


class GroundTruthTenderSchema(BaseModel):
    id: str
    bidNumber: str
    title: str
    buyer: str
    category: str
    estimatedValue: int
    estimatedValueFormatted: str
    emdAmount: int
    emdAmountFormatted: str
    submissionDeadline: str
    description: str
    status: str


class TenderRequirementSchema(BaseModel):
    id: str
    tenderId: str
    requirementCode: str
    category: str
    requirementText: str
    requirementType: str
    operator: str
    thresholdValue: Optional[Any] = None
    thresholdUnit: Optional[str] = None
    mandatory: bool
    exemptionRule: Optional[str] = None
    sourceDocument: str
    sourcePage: int
    sourceClause: str


class GroundTruthBidderSchema(BaseModel):
    id: str
    tenderId: str
    bidderCode: str
    legalName: str
    shortName: str
    cin: str
    pan: str
    gstin: str
    udyamNumber: str
    enterpriseType: str
    businessType: str
    registeredAddress: str
    state: str
    city: str
    pincode: str
    authorizedSignatory: str
    phone: str
    email: str


class BidderDocumentSchema(BaseModel):
    id: str
    bidderId: str
    documentCode: str
    documentName: str
    documentType: str
    pageNumber: int
    documentDate: Optional[str] = None
    validUntil: Optional[str] = None
    documentStatus: str
    fileName: str


class BidderEvidenceSchema(BaseModel):
    id: str
    bidderId: str
    documentId: str
    evidenceType: str
    fieldName: str
    extractedValue: str
    normalizedValue: Optional[Any] = None
    unit: Optional[str] = None
    confidence: float
    sourcePage: int
    sourceText: str


class ComplianceResultSchema(BaseModel):
    id: str
    bidderId: str
    requirementId: str
    status: str
    submittedValue: str
    requiredValue: str
    reason: str
    evidenceId: str
    confidence: float
    reviewRequired: bool


class BidderBenchmarkSchema(BaseModel):
    bidderId: str
    bidderCode: str
    legalName: str
    tenderId: str
    bidNumber: str
    benchmarkStatus: str
    benchmarkLabel: str
    summaryReason: str
    failingRequirementsCount: int
    reviewRequirementsCount: int
    passingRequirementsCount: int


class GroundTruthDatasetResponse(BaseModel):
    tenders: list[GroundTruthTenderSchema]
    requirements: list[TenderRequirementSchema]
    bidders: list[GroundTruthBidderSchema]
    documents: list[BidderDocumentSchema]
    evidence: list[BidderEvidenceSchema]
    complianceResults: list[ComplianceResultSchema]
    benchmarks: list[BidderBenchmarkSchema]

File: src/types/ground-truth.ts
// ============================================================
// SatyaSetu — Ground Truth Data Layer Types (Phase 3)
// Canonical Tender Requirements, Bidder Submissions & Evidence
// ============================================================

export type RequirementCategory =
  | 'MANDATORY'
  | 'TECHNICAL'
  | 'FINANCIAL'
  | 'STATUTORY'
  | 'EXPERIENCE'
  | 'LOCATION'
  | 'PREFERENTIAL'
  | 'COMMERCIAL';

export type RequirementType =
  | 'NUMERIC'
  | 'BOOLEAN'
  | 'DOCUMENT'
  | 'VALIDITY'
  | 'EXPERIENCE'
  | 'LOCATION'
  | 'TECHNICAL'
  | 'FINANCIAL'
  | 'STATUTORY'
  | 'EXEMPTION'
  | 'DECLARATION';

export type RequirementOperator =
  | '>='
  | '<='
  | '=='
  | 'EXISTS'
  | 'VALID'
  | 'MATCH'
  | 'CONTAINS'
  | 'NOT_EXISTS';

export interface GroundTruthTender {
  id: string;
  bidNumber: string;
  title: string;
  buyer: string;
  category: string;
  estimatedValue: number;
  estimatedValueFormatted: string;
  emdAmount: number;
  emdAmountFormatted: string;
  submissionDeadline: string;
  publishDate: string;
  bidValidityDays: number;
  evaluationType: string;
  deliveryLocation: string;
  deliveryPeriodDays: number;
  warrantyMonths: number;
  description: string;
  status: 'OPEN' | 'EVALUATION' | 'CLOSED' | 'AWARDED';
}

export interface TenderRequirement {
  id: string;
  tenderId: string;
  requirementCode: string;
  category: RequirementCategory;
  requirementText: string;
  requirementType: RequirementType;
  operator: RequirementOperator;
  thresholdValue?: string | number | boolean;
  thresholdUnit?: string;
  mandatory: boolean;
  exemptionRule?: string;
  sourceDocument: string;
  sourcePage: number;
  sourceClause: string;
}

export type EnterpriseType = 'MICRO' | 'SMALL' | 'MEDIUM' | 'LARGE';
export type BusinessType = 'MANUFACTURER' | 'SERVICE_PROVIDER' | 'TRADER' | 'RESELLER' | 'CONTRACTOR';

export interface GroundTruthBidder {
  id: string;
  tenderId: string;
  bidderCode: string; // e.g. T1-B2, T2-B1, T2-B2, T3-B1, T3-B2
  legalName: string;
  shortName: string;
  cin: string;
  pan: string;
  gstin: string;
  udyamNumber: string;
  enterpriseType: EnterpriseType;
  businessType: BusinessType;
  registeredAddress: string;
  state: string;
  city: string;
  pincode: string;
  authorizedSignatory: string;
  designation: string;
  phone: string;
  email: string;
  incorporationYear: number;
}

export type GroundTruthDocumentType =
  | 'PAN'
  | 'GST'
  | 'UDYAM'
  | 'TURNOVER_CERTIFICATE'
  | 'EXPERIENCE_CERTIFICATE'
  | 'PURCHASE_ORDER'
  | 'INVOICE'
  | 'MAF'
  | 'MII_DECLARATION'
  | 'ELECTRICAL_LICENSE'
  | 'GST_RETURN'
  | 'NON_BLACKLISTING'
  | 'SOLVENCY'
  | 'TECHNICAL_COMPLIANCE'
  | 'ADDRESS_PROOF'
  | 'EMD_EXEMPTION'
  | 'STARTUP_CERTIFICATE'
  | 'SAMPLE_UNDERTAKING'
  | 'OTHER';

export type DocumentStatus = 'VALID' | 'EXPIRED' | 'DEFICIENT' | 'PROVISIONAL' | 'UNNOTARIZED' | 'MISSING';

export interface BidderDocument {
  id: string;
  bidderId: string;
  documentCode: string;
  documentName: string;
  documentType: GroundTruthDocumentType;
  pageNumber: number;
  documentDate?: string;
  validFrom?: string;
  validUntil?: string;
  documentStatus: DocumentStatus;
  fileName: string;
}

export type EvidenceType =
  | 'TURNOVER'
  | 'EXPERIENCE'
  | 'AUTHORIZATION'
  | 'STATUTORY_ID'
  | 'LOCATION'
  | 'MII'
  | 'EMD'
  | 'LICENSE'
  | 'TAX_FILING'
  | 'DECLARATION'
  | 'TECHNICAL_SPEC'
  | 'FINANCIAL_STANDING';

export interface BidderEvidence {
  id: string;
  bidderId: string;
  documentId: string;
  evidenceType: EvidenceType;
  fieldName: string;
  extractedValue: string;
  normalizedValue: string | number | boolean | null;
  unit?: string;
  confidence: number;
  sourcePage: number;
  sourceText: string;
}

export type GroundTruthComplianceStatus =
  | 'PASS'
  | 'FAIL'
  | 'REVIEW'
  | 'NOT_APPLICABLE'
  | 'PENDING';

export interface GroundTruthComplianceResult {
  id: string;
  bidderId: string;
  requirementId: string;
  status: GroundTruthComplianceStatus;
  submittedValue: string;
  requiredValue: string;
  reason: string;
  evidenceId: string;
  confidence: number;
  reviewRequired: boolean;
}

export type BenchmarkStatus = 'COMPLIANT' | 'NON_COMPLIANT' | 'REVIEW';

export interface BidderBenchmark {
  bidderId: string;
  bidderCode: string;
  legalName: string;
  tenderId: string;
  bidNumber: string;
  benchmarkStatus: BenchmarkStatus;
  benchmarkLabel: string;
  summaryReason: string;
  failingRequirementsCount: number;
  reviewRequirementsCount: number;
  passingRequirementsCount: number;
}
